### PR TITLE
[Microservice] Security Group Manager, Port Manager Refactor

### PR DESCRIPTION
**Pull request by chenpiaoping**
This PR contains the following features:

* Security Group Manager
-- Security group create/update/delete/query interfaces
-- Security group rule create/delete/query interfaces
-- Port bind/unbind security group interfaces

* Ignite
-- Add a thin Ignite Client to support SQL-alike query

* Port Manager
-- Add port data modeling and new dao layer to support SQL-alike query

### CODE REVIEW

File: lib/src/main/java/com/futurewei/alcor/common/utils/Ipv4AddrUtil.java [hunk 1]
```diff
@@ -63,4 +68,13 @@ public static String longToIpv4(long longIpv4) {
 
         return result;
     }
+
+    public static boolean ipv4PrefixCheck(String strIpv4) {
```
Comment (xieus) [inline lib/src/main/java/com/futurewei/alcor/common/utils/Ipv4AddrUtil.java:72]: Like it!

Comment (xieus) [reply lib/src/main/java/com/futurewei/alcor/common/utils/Ipv4AddrUtil.java:72]: Let us get some UTs to cover this basic util method.

File: services/security_group_manager/pom.xml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        <java.version>1.8</java.version>
+    </properties>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-web</artifactId>
+        </dependency>
+
+        <dependency>
```
Comment (xieus) [inline services/security_group_manager/pom.xml:27]: This pom is extremely concise! Very good.

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/exception/DefaultSecurityGroupExists.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        See the License for the specific language governing permissions and
+        limitations under the License.
+*/
+package com.futurewei.alcor.securitygroup.exception;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(code= HttpStatus.PRECONDITION_FAILED, reason="Default security group exists")
+public class DefaultSecurityGroupExists extends Exception {
```
Comment (xieus) [inline services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/exception/DefaultSecurityGroupExists.java:22]: I think we should have a good design doc and include details of all these good exception.

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/repo/SecurityGroupBindingsRepository.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    private void init() {
+        LOG.info("SecurityGroupBindingsRepository init done");
+    }
+
+    public SecurityGroupBindings getSecurityGroupBindings(String id) throws CacheException {
+        return bindingCache.get(id);
+    }
+
+
+    public void addSecurityGroupBinding(PortSecurityGroupsJson portSecurityGroupsJson) throws Exception {
```
Comment (xieus) [inline services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/repo/SecurityGroupBindingsRepository.java:58]: Could we add binding to a subnet?

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/service/implement/SecurityGroupServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        }
+
+        LOG.info("List security group success");
+
+        return securityGroups;
+    }
+
+    @Override
+    public PortSecurityGroupsJson bindSecurityGroups(PortSecurityGroupsJson portSecurityGroupsJson) throws Exception {
+        securityGroupBindingsRepository.addSecurityGroupBinding(portSecurityGroupsJson);
```
Comment (xieus) [inline services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/service/implement/SecurityGroupServiceImpl.java:227]: For binding, we should also update port manager, right?

File: lib/src/main/java/com/futurewei/alcor/common/utils/Ipv4AddrUtil.java (hunk 1 above)
Comment (chenpiaoping) [reply lib/src/main/java/com/futurewei/alcor/common/utils/Ipv4AddrUtil.java:72]: no problem

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/exception/DefaultSecurityGroupExists.java (hunk 3 above)
Comment (chenpiaoping) [reply services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/exception/DefaultSecurityGroupExists.java:22]: no problem

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/repo/SecurityGroupBindingsRepository.java (hunk 4 above)
Comment (chenpiaoping) [reply services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/repo/SecurityGroupBindingsRepository.java:58]: No at present, are we going to support that?

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/service/implement/SecurityGroupServiceImpl.java (hunk 5 above)
Comment (chenpiaoping) [reply services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/service/implement/SecurityGroupServiceImpl.java:227]: yes, I will do that after I finish the batch interfaces of security group.

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/repo/SecurityGroupBindingsRepository.java (hunk 4 above)
Comment (xieus) [reply services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/repo/SecurityGroupBindingsRepository.java:58]: Hold on for a while here. We could use introduce NACL on subnet level and keep security group on port level.

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupController.java [hunk 6]
```diff
@@ -31,67 +34,129 @@
...
     @ResponseStatus(HttpStatus.CREATED)
     public SecurityGroupJson createSecurityGroup(@PathVariable("project_id") String projectId,
                                              @RequestBody SecurityGroupJson securityGroupJson) throws Exception {
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(projectId);
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(securityGroupJson);
-        SecurityGroup securityGroup = securityGroupJson.getSecurityGroup();
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(securityGroup);
-        RestPreconditionsUtil.verifyParameterNotNullorEmpty(securityGroup.getTenantId());
-        securityGroup.setProjectId(projectId);
+        checkProjectId(projectId);
```
Comment (xieus) [inline services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupController.java:86]: In the future, we could write a general microservice validation framework, which supports an abstraction of validator and includes a list of abstract rules. A validator implementation for a specific microservice could include specific set of rules, for example, validateProjectId, validateSecurityGroup etc.

File: docs/design/security_group_manager.adoc [hunk 7]
```diff
@@ -0,0 +1,1132 @@
...
+. Managing security group and security group rule resources.
+. Create (port, Security Group) association when a port is created.
+. Release (port, Security Group) association when a port is released.
+. Concurrency control mechanism should be provided to process multiple concurrent bind/unbind Security Group requests.
+. Working together with other services including subnet manager, port manager.
+
+
+== REST APIs
+
+=== API Snapshot
```
Comment (xieus) [inline docs/design/security_group_manager.adoc:27]: SGM indeed has a lot of APIs. Give me a bit time to review. Good job in writing them down.

**Comment by xieus** [comment]
@chenpiaoping I am on this PR now.

File: services/port_manager/src/main/java/com/futurewei/alcor/portmanager/util/GoalStateUtil.java [hunk 8]
```diff
@@ -270,9 +273,9 @@ private static int protocolToNumber(String protocol) {
             } else if (entity instanceof SubnetEntity) {
                 SubnetEntity subnetEntity = (SubnetEntity) entity;
                 subnetEntityMap.put(subnetEntity.getId(), subnetEntity);
-            } else if (entity instanceof SecurityGroupEntity) {
-                SecurityGroupEntity securityGroupEntity = (SecurityGroupEntity) entity;
-                securityGroupEntityMap.put(securityGroupEntity.getId(), securityGroupEntity);
+            } else if (entity instanceof SecurityGroup) {
+                SecurityGroup securityGroupEntity = (SecurityGroup) entity;
```
Comment (xieus) [inline services/port_manager/src/main/java/com/futurewei/alcor/portmanager/util/GoalStateUtil.java:277]: Seems that the latest Alcor/master is not merge yet. We should have SecurityGroupEntity class here @chenpiaoping

Comment (chenpiaoping) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/util/GoalStateUtil.java:277]: I renamed it SecurityGroup again.

Comment (xieus) [reply services/port_manager/src/main/java/com/futurewei/alcor/portmanager/util/GoalStateUtil.java:277]: Okay it is fine.

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+@RestController
+public class SecurityGroupRuleController {
+    @Autowired
+    private SecurityGroupRuleService securityGroupRuleService;
+
+    private void checkProjectId(String projectId) throws ProjectIdRequired {
```
Comment (xieus) [inline services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java:38]: For the private check methods, I would recommend to move to an util class which could be replaced by a validator.

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+        String etherType = securityGroupRule.getEtherType();
+
+        List<String> ipv6Protocols = Arrays.asList("icmpv6");
+        if (ipv6Protocols.contains(protocol) &&
+                SecurityGroupRule.EtherType.IPV4.getType().equals(etherType)) {
+            throw new ProtocolEtherTypeConflict();
+        }
+    }
+
+    private void checkSecurityGroupRule(SecurityGroupRule securityGroupRule) throws Exception {
```
Comment (xieus) [inline services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java:157]: Didn't see protocol check here. Which protocols are supported by the v1 implementation. Could you list the current scope at the beginning of the controller.

File: web/src/main/java/com/futurewei/alcor/web/entity/securitygroup/SecurityGroupRule.java [hunk 11]
```diff
@@ -15,7 +18,7 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.futurewei.alcor.common.entity.CustomerResource;
 
-public class SecurityGroupRuleEntity extends CustomerResource {
+public class SecurityGroupRule extends CustomerResource {
```
Comment (xieus) [inline web/src/main/java/com/futurewei/alcor/web/entity/securitygroup/SecurityGroupRule.java:21]: Compared with Neutron SG rule, we need three more fields:

revision_number, created_at, updated_at.

Consider to add it into CustomerResource.

File: docs/modules/ROOT/pages/mgmt_services/security_group_manager.adoc [hunk 12]
```diff
@@ -0,0 +1,1132 @@
...
+Response:
+{
+	"security_group_rule": {
+		"project_id": "3d53801c-32ce-4e97-9572-bb966fab628",
+		"id": "3d53801c-32ce-4e97-9572-bb966fa6b23",
+		"name": null,
+		"description": null,
+		"tenant_id": "3d53801c-32ce-4e97-9572-bb966fab628",
+		"security_group_id": "3d53801c-32ce-4e97-9572-bb966f476ec",
+		"remote_group_id": null,
```
Comment (xieus) [inline docs/modules/ROOT/pages/mgmt_services/security_group_manager.adoc:633]: remote_group_id and remote_group_prefix can't be null at the same time. Please check if we have appropriate check in the code.

Comment (chenpiaoping) [reply docs/modules/ROOT/pages/mgmt_services/security_group_manager.adoc:633]: yes, we have.

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java (hunk 10 above)
Comment (chenpiaoping) [reply services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java:157]: Oh, let check.

File: web/src/main/java/com/futurewei/alcor/web/entity/securitygroup/SecurityGroupRule.java (hunk 11 above)
Comment (chenpiaoping) [reply web/src/main/java/com/futurewei/alcor/web/entity/securitygroup/SecurityGroupRule.java:21]: ok

File: services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java (hunk 9 above)
Comment (chenpiaoping) [reply services/security_group_manager/src/main/java/com/futurewei/alcor/securitygroup/controller/SecurityGroupRuleController.java:38]: ok

File: docs/modules/ROOT/pages/mgmt_services/security_group_manager.adoc (hunk 12 above)
Comment (xieus) [reply docs/modules/ROOT/pages/mgmt_services/security_group_manager.adoc:633]: In this case, can we use update the sample response? We have both fields null here.

File: services/port_manager/src/test/java/com/futurewei/alcor/portmanager/entity/Address.java [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+import org.apache.ignite.binary.Binarylizable;
+
+/**
+ * Employee address.
+ * <p>
+ * This class implements {@link Binarylizable} only for example purposes,
+ * in order to show how to customize serialization and deserialization of
+ * binary objects.
+ */
+public class Address implements Binarylizable {
```
Comment (xieus) [inline services/port_manager/src/test/java/com/futurewei/alcor/portmanager/entity/Address.java:32]: Not Alcor classes. Please remove.

File: web/src/main/java/com/futurewei/alcor/web/entity/securitygroup/SecurityGroupRule.java (hunk 11 above)
Comment (xieus) [reply web/src/main/java/com/futurewei/alcor/web/entity/securitygroup/SecurityGroupRule.java:21]: Tracked by Issue https://github.com/futurewei-cloud/alcor/issues/246.

File: services/port_manager/src/test/java/com/futurewei/alcor/portmanager/entity/Address.java (hunk 13 above)
Comment (chenpiaoping) [reply services/port_manager/src/test/java/com/futurewei/alcor/portmanager/entity/Address.java]: These files have been deleted

**Pull request merged**